File: algorithms/merge_sort.cpp
#include <iostream>
#include <stdlib.h>
#include <time.h>
 
int ARRAY_SIZE = 10;
 
void insertion_sort(int *arr, int size) {
  for ( int i = 0; i < size; i++ ) {
    int key = arr[i];
    int j = i - 1;
    while (( j >= 0) && key < arr[j]) {
      arr[j+1] = arr[j];
      j--;
    }
    arr[j+1] = key;
  }
}

int* merge_sort(int *arr, int size) {
  if (size == 1) {
    return arr;
  } else {
    
    // setup for the merging
    int arr_size_0 = 0;
    int arr_size_1 = 0;
    
    if (size % 2 == 0) {
      arr_size_0 = arr_size_1 = size / 2;
    } else {
      arr_size_0 = (size / 2) + 1;
      arr_size_1 = (size / 2);
    }
    
    int arr0[arr_size_0];
    for ( int i = 0; i < arr_size_0; i++ ) {
      arr0[i] = arr[i];
    }
    
    int arr1[arr_size_1];
    for ( int i = 0; i < arr_size_1; i++ ){
      arr1[i] = arr[arr_size_1 + i];
    }
    
    int *arr0_sorted = merge_sort(arr0, arr_size_0);
    int *arr1_sorted = merge_sort(arr1, arr_size_1);
    
    int i0, i1, i = 0;
    while ((i0 < arr_size_0) || (i1 < arr_size_1)) {
      if (i0 >= arr_size_0) {
        arr[i] = arr1_sorted[i1];
        i++;
        i1++;
      } else if ( i1 >= arr_size_1 ) {
        arr[i] = arr0_sorted[i0];
        i++;
        i0++;
      } else {
        if (arr0_sorted[i0] < arr1_sorted[i1]) {
          arr[i] = arr0_sorted[i0];
          i0++;
          i++;
        } else {
          arr[i] = arr1_sorted[i1];
          i1++;
          i++;
        }
      }
    }
    
    for (int i = 0; i < ARRAY_SIZE; i++ ) {
      std::cout << arr_sorted[i] << ' ';
    }
    std:: cout << "\n\n";
    return arr;
    
  }
}
 
int main() {
  srand(time(NULL));
  int arr[ARRAY_SIZE];
  for (int num = 0; num < ARRAY_SIZE; num++ ) {
    arr[num] = rand() % ARRAY_SIZE;
  }
  
  std::cout << "Original sequence: \n";
  for (int i = 0; i < ARRAY_SIZE; i++ ) {
    std::cout << arr[i] << ' ';
  }
  std:: cout << "\n\n";
  
  int *arr_sorted = merge_sort(arr, ARRAY_SIZE);
  std::cout << "Sorted sequence: \n";
  for (int i = 0; i < ARRAY_SIZE; i++ ) {
    std::cout << arr_sorted[i] << ' ';
  }
  std:: cout << "\n\n";
  
}

File: algorithms/merge_sort.py
import random

def merge_sort(arr):
    if len(arr) == 1:
        return arr
    else:
        arr1 = merge_sort(arr[0:len(arr) // 2])
        arr2 = merge_sort(arr[len(arr) // 2: len(arr)])

        il = i2 = i = 0
        while il < len(arr1) or i2 < len(arr2):
            if il >= len(arr1):
                arr[i] = arr2[i2]
                i += 1
                i2 += 1
            elif i2 >= len(arr2):
                arr[i] = arr1[il]
                i += 1
                il += 1
            else:
                if arr1[il] < arr2[i2]:
                    arr[i] = arr1[il]
                    il += 1
                    i += 1
                else:
                    arr[i] = arr2[i2]
                    i2 += 1
                    i += 1
        return arr

if __name__ == '__main__':
    test_arrs = [[random.randint(0, 1000) for i in range(1, 21)] for num in range(1, 21)] 

    for i, test_arr in enumerate(test_arrs):
        print("Orignal {}: ".format(i), test_arr)
        print("Sorted {}: ".format(i), merge_sort(test_arr))